### PR TITLE
Streamline RSpec config

### PR DESCRIPTION
**Pull request by pirj**
Disable monkey patching and fix specs related to this.

### CODE REVIEW

File: spec/rspec/rails/example/view_example_group_spec.rb [hunk 1]
```diff
@@ -58,7 +58,7 @@ module ::ApplicationHelper; end
         end
 
         it 'includes the application helper' do
-          group = RSpec::Core::Example.describe 'bars/new.html.erb'
+          group = RSpec::Core::ExampleGroup.describe 'bars/new.html.erb'
```
Comment (pirj) [inline spec/rspec/rails/example/view_example_group_spec.rb]: This is interesting.

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -25,17 +26,34 @@ def self.run_all(reporter=nil)
...
 end
 
+# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
 RSpec.configure do |config|
+  config.expect_with :rspec do |expectations|
+    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
+  end
+
+  config.mock_with :rspec do |mocks|
+    # mocks.verify_partial_doubles = true
```
Comment (pirj) [inline spec/spec_helper.rb]: A lot of specs fail. Not sure if it's worth it considering we're only using `instance_double` once.

Comment (pirj) [reply spec/spec_helper.rb]: Complete nonsense ☝️

Comment (pirj) [reply spec/spec_helper.rb]: A lot of specs fail, e.g.:
```
  17) RSpec::Rails::SetupAndTeardownAdapter::setup registers teardown hooks in the order setup is received
      Failure/Error: expect(klass).to receive(:after).ordered { |&block| expect(block.call).to eq "foo" }
        #<Class:0x00007fdb8f30ec70> does not implement: after
      # ./spec/rspec/rails/setup_and_teardown_adapter_spec.rb:40:in `block (3 levels) in <top (required)>'

  18) render_template given a symbol converts to_s and delegates to assert_template
      Failure/Error: expect(self).to receive(:assert_template).with("template_name", "this message")
        #<RSpec::ExampleGroups::RenderTemplate::GivenASymbol "converts to_s and delegates to assert_template" (./spec/rspec/rails/matchers/have_rendered_spec.rb:23)> does not implement: assert_template
```

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 1 above)
Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: I mean... I'm not sure theres a difference :joy:

File: spec/rspec/rails/matchers/action_cable/have_broadcasted_to_spec.rb [hunk 3]
```diff
@@ -19,7 +19,7 @@ def to_global_id(options = {})
   end
 end
 
-RSpec.describe "have_broadcasted_to matchers", :skip => !RSpec::Rails::FeatureCheck.has_action_cable_testing? do
+RSpec.describe "have_broadcasted_to matchers", :if => RSpec::Rails::FeatureCheck.has_action_cable_testing? do
```
Comment (JonRowe) [inline spec/rspec/rails/matchers/action_cable/have_broadcasted_to_spec.rb]: This behaves differently, `if` means it won't have the `skipped` message?

File: spec/spec_helper.rb (hunk 2 above)
Comment (JonRowe) [reply spec/spec_helper.rb]: There is a method that can surpress verification for specific spec's that need it.

File: spec/rspec/rails/example/controller_example_group_spec.rb [hunk 4]
```diff
@@ -84,7 +83,6 @@ def my_helper
 
       it "calls NamedRouteCollection#route_defined? when it checks that given route is defined or not" do
         expect(routes.named_routes).to receive(:route_defined?).and_return(true)
-        expect(routes.named_routes).not_to receive(:helpers)
```
Comment (pirj) [inline spec/rspec/rails/example/controller_example_group_spec.rb]: Frankly, I'm not sure what it has to do with helpers.

File: spec/rspec/rails/example/view_example_group_spec.rb [hunk 5]
```diff
@@ -225,6 +235,7 @@ def controller
 
       it 'is accessible to hooks' do
         with_isolated_config do
+          view = double("view")
```
Comment (pirj) [inline spec/rspec/rails/example/view_example_group_spec.rb]: No idea why it was passing before.

File: spec/rspec/rails/example/view_example_group_spec.rb [hunk 6]
```diff
@@ -124,6 +130,7 @@ def _assigns
             end
           end
           include local
```
Comment (pirj) [inline spec/rspec/rails/example/view_example_group_spec.rb]: Quite surprised, but defining those methods direcly in the class breaks the spec.

File: spec/rspec/rails/example/controller_example_group_spec.rb (hunk 4 above)
Comment (JonRowe) [reply spec/rspec/rails/example/controller_example_group_spec.rb]: Check the original history with blame, I imagine this is a regression check of some kind

File: spec/rspec/rails/matchers/have_rendered_spec.rb [hunk 7]
```diff
@@ -1,8 +1,10 @@
 %w[have_rendered render_template].each do |template_expectation|
-  describe template_expectation do
+  RSpec.describe template_expectation do
     include RSpec::Rails::Matchers::RenderTemplate
     let(:response) { ActionDispatch::TestResponse.new }
 
+    def assert_template(*); end # Stub method
```
Comment (JonRowe) [inline spec/rspec/rails/matchers/have_rendered_spec.rb]: Why did you move this?

Comment (pirj) [reply spec/rspec/rails/matchers/have_rendered_spec.rb]: Not only that example that I've moved it from need `assert_template` to be defined on `self`.
I was on the fence of keeping it there for clarity. Should I?

File: spec/rspec/rails/example/controller_example_group_spec.rb (hunk 4 above)
Comment (pirj) [reply spec/rspec/rails/example/controller_example_group_spec.rb]: This needs some deeper research on how it works with 4.2 and later.
I'll comment this out instead and add a todo, there might be a potential to get rid of some legacy in `controller_example_group.rb`.

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 6 above)
Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: I'm not sure what this refers to

File: spec/rspec/rails/example/view_example_group_spec.rb [hunk 8]
```diff
@@ -189,8 +196,10 @@ def controller
       let(:view_spec) do
         Class.new do
           include ViewExampleGroup::ExampleMethods
+          def _default_file_to_render; end
```
Comment (JonRowe) [inline spec/rspec/rails/example/view_example_group_spec.rb]: What needs this?

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 5 above)
Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: Is `view` not also a method?

File: spec/rspec/rails/example/view_example_group_spec.rb [hunk 9]
```diff
@@ -213,6 +222,7 @@ def controller
     describe "#view" do
       let(:view_spec) do
         Class.new do
+          def _view; end # Stub method
```
Comment (JonRowe) [inline spec/rspec/rails/example/view_example_group_spec.rb]: Whats this for?

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 6 above)
Comment (pirj) [reply spec/rspec/rails/example/view_example_group_spec.rb]: I've tried to:
```ruby
Class.new do
  def received
    @received ||= []
  end
  # ...
```
but this doesn't work.

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 8 above)
Comment (pirj) [reply spec/rspec/rails/example/view_example_group_spec.rb]: [This](https://github.com/rspec/rspec-rails/pull/2264/files/87aee32e014aed4c818156a9c473c5de8682b8cc#diff-0cdf276129b0b83c6314283584c4df6eR205) and [this](https://github.com/rspec/rspec-rails/pull/2264/files/87aee32e014aed4c818156a9c473c5de8682b8cc#diff-0cdf276129b0b83c6314283584c4df6eR214).

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 9 above)
Comment (pirj) [reply spec/rspec/rails/example/view_example_group_spec.rb]: It's to satisfy [this](https://github.com/rspec/rspec-rails/pull/2264/files/87aee32e014aed4c818156a9c473c5de8682b8cc#diff-0cdf276129b0b83c6314283584c4df6eR232).

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 5 above)
Comment (pirj) [reply spec/rspec/rails/example/view_example_group_spec.rb]: Doesn't seem so, at least not it the bare context of `ExampleGroup.describe`.
This is needed [here](https://github.com/rspec/rspec-rails/pull/2264/files/87aee32e014aed4c818156a9c473c5de8682b8cc#diff-0cdf276129b0b83c6314283584c4df6eR241).
Picked it up [here](https://github.com/rspec/rspec-rails/pull/2264/files/87aee32e014aed4c818156a9c473c5de8682b8cc#diff-0cdf276129b0b83c6314283584c4df6eR231).
I suspect that `with_isolated_config` was swallowing `NoMethodError` previously.

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 6 above)
Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: Doesn't work as in?

File: spec/rspec/rails/example/controller_example_group_spec.rb (hunk 4 above)
Comment (JonRowe) [reply spec/rspec/rails/example/controller_example_group_spec.rb]: I'd not comment it out, but feel free to leave a todo to investigate.

File: spec/rspec/rails/matchers/have_rendered_spec.rb (hunk 7 above)
Comment (JonRowe) [reply spec/rspec/rails/matchers/have_rendered_spec.rb]: It makes it available to more specs, which is not necessarily desired.

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 8 above)
Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: Aside, can you quote code as well as link, Github makes it really hard to navigate with just links...

Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: Ok I see it now, it's just a bit weird because its not in the same context.

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 9 above)
Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: This link doesn't seem to work

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 5 above)
Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: Nope nor these links

Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: `with_isolated_config` does not swallow errors, it just runs an `ensure`

Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: It's defined here: `lib/rspec/rails/example/view_example_group.rb:81` so this shouldn't be removed...

File: spec/rspec/rails/example/view_example_group_spec.rb [hunk 10]
```diff
@@ -225,6 +235,7 @@ def controller
 
       it 'is accessible to hooks' do
         with_isolated_config do
+          view = double("view")
           run_count = 0
           RSpec.configuration.before(:each, :type => :view) do
             allow(view).to receive(:a_stubbed_helper) { :value }
```
Comment (JonRowe) [inline spec/rspec/rails/example/view_example_group_spec.rb]: This runs in a `ViewExampleGroup` where `view` is a method.

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 6 above)
Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: No I meant how doesn't it work, how does:
```
Class.new do
  def received
    @received ||= []
  end
end
```
Not have the same effect as above? What is the error it causes.

Comment (pirj) [reply spec/rspec/rails/example/view_example_group_spec.rb]: It's this:
```
  4) RSpec::Rails::ViewExampleGroup#render given no input sends render(:template => (described file)) to the view
     Failure/Error: expect(view_spec.received.first).to eq([{:template => "widgets/new"},{}, nil])

       expected: [{:template=>"widgets/new"}, {}, nil]
            got: [{}, {}, nil]

       (compared using ==)

       Diff:
       @@ -1,2 +1,2 @@
       -[{:template=>"widgets/new"}, {}, nil]
       +[{}, {}, nil]
```

It does not make sense to me, I don't see any `prepend` coming into play.

File: spec/rspec/rails/example/view_example_group_spec.rb [hunk 11]
```diff
@@ -213,6 +222,7 @@ def controller
     describe "#view" do
       let(:view_spec) do
```
[the remark this answers is not in the document]
Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: Nah

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 5 above)
Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: You need to keep `with_isolated_config`, if your changes here are causing an error you need to ensure the right methods are defined, doing `double` here is just wrong..

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 6 above)
Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: It must be something similar, as this exists as is I suggest leaving it.

File: spec/rspec/rails/example/view_example_group_spec.rb (hunk 5 above)
Comment (pirj) [reply spec/rspec/rails/example/view_example_group_spec.rb]: A call to `allow(view).to receive(:a_stubbed_helper) { :value }` results in:
```
RSpec::Mocks::MockExpectationError: #<#<Class:0x00007fa59165da18>:0x00007fa59165ceb0 @_config={}, ...> does not implement: a_stubbed_helper
```
that for some reason doesn't bubble up.
The same happens if you move stubbing directly to the example:
```ruby
          group = RSpec::Core::ExampleGroup.describe 'a view', :type => :view do
            specify {
              allow(view).to receive(:a_stubbed_helper) { :value }
              run_count += 1
            }
          end
```
Example returns prematurely, `run_count` is not incremented, but there's no exception from mocks. Is it a bug?

Changed to stub an existing `render` method instead.

Removed `NullObject.new` from `group.run`'s arguments, it doesn't seem to affect this in any way.

**Comment by pirj** [comment]
Rebase fixed, all green.

File: spec/rspec/rails/example/view_example_group_spec.rb [hunk 12]
```diff
@@ -227,13 +237,13 @@ def controller
...
           RSpec.configuration.before(:each, :type => :view) do
-            allow(view).to receive(:a_stubbed_helper) { :value }
+            allow(view).to receive(:render) { :value }
             run_count += 1
           end
           group = RSpec::Core::ExampleGroup.describe 'a view', :type => :view do
             specify { true }
           end
-          group.run NullObject.new
+          group.run
```
Comment (JonRowe) [inline spec/rspec/rails/example/view_example_group_spec.rb]: Doesn't this cause the groups output to be output to the real formatter?

Comment (pirj) [reply spec/rspec/rails/example/view_example_group_spec.rb]: There's no difference in the output.

Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: I was surprised by that assertion, as the point of this object is to prevent the real reporter being used, but its failing locally for me? I want to investigate this further as its not possible to prove either way on CI without changing the format...

What I expect this to do is to add the spec you've specified above into the reporter, so if you run this spec on its own you'd see "2 examples passed" for this one example.

Comment (pirj) [reply spec/rspec/rails/example/view_example_group_spec.rb]: If I run it focused (`rspec spec/rspec/rails/example/view_example_group_spec.rb:236`), it fails locally. However, the spec when run altogether passes.
This weird behavior remains even with just one example left, and if you remove `module` namespace, and `with_isolated_`, and nesting contexts. I've put it on my list to investigate.

Comment (pirj) [reply spec/rspec/rails/example/view_example_group_spec.rb]: Maybe it doesn't add to the main reporter because of https://github.com/rspec/rspec-rails/pull/2264/files#diff-93830fa29d616f7c87903d08b5b1b29aR55 ?
```ruby
  config.around(:example) do |example|
    real_world = RSpec.world
    RSpec.instance_variable_set(:@world, RSpec::Core::World.new)
    example.run
    RSpec.instance_variable_set(:@world, real_world)
  end
```

Comment (pirj) [reply spec/rspec/rails/example/view_example_group_spec.rb]: I'm not sure if `with_isolated_config` is even necessary in specs with that hook.

Comment (JonRowe) [reply spec/rspec/rails/example/view_example_group_spec.rb]: So I got to the bottom of why this doesn't run when focused. If you're interested its an artefact in our filtering, `with_isolated_config` does not remove filters from this spec, as they exist in the world and the world is not isolated. That is something I might consider fixing at some point but is waaaaaaaaay out of scope for this.

I also confirmed this removal is ok, we now use a `NullReporter` by default, a change which post-dates this spec I suspect.
